packages-exp/auth-exp/src/core/user/token_manager.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IdTokenResponse } from '../../model/id_token';
+
+/**
+ * The number of milliseconds before the official expiration time of a token
+ * to refresh that token, to provide a buffer for RPCs to complete.
+ */
+const TOKEN_REFRESH_BUFFER_MS = 30_000;
+
+export interface Tokens {
+  accessToken: string;
+  refreshToken: string | null;
+}
+
+export const DATE_GENERATOR = {
+  now: () => Date.now()
+};
+
+export class StsTokenManager {
+  refreshToken: string | null = null;
```

Comment:
do we have a standard pattern on when we use | null vs foo? ?

Reply:
We don't but we should settle on one. Either is acceptable but we should be consistent.

I personally prefer null when a field _should always be present_ but *can* be null. There is a semantic difference between:
`const a = {b: undefined}`
and `const a = {}`

Reply:
I like that, in that case we should probably use `| null` in most places